src/libPMacc/include/particles/operations/CountParticles.hpp

```diff
@@ -31,72 +30,163 @@
 
-        typedef typename Mapping::SuperCellSize SuperCellSize;
+        DataSpace< dim > const threadIndex( threadIdx );
+        uint32_t const workerIdx = static_cast< uint32_t >(
+            DataSpaceOperations< dim >::template map< SuperCellSize >( threadIndex )
+        );
 
-        const DataSpace<Dim > threadIndex(threadIdx);
-        const int linearThreadIdx = DataSpaceOperations<Dim>::template map<SuperCellSize > (threadIndex);
-        const DataSpace<Dim> superCellIdx(mapper.getSuperCellIndex(DataSpace<Dim > (blockIdx)));
+        DataSpace< dim > const superCellIdx( mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) ) );
 
-        if (linearThreadIdx == 0)
-        {
-            frame = pb.getLastFrame(superCellIdx);
-            particlesInSuperCell = pb.getSuperCell(superCellIdx).getSizeLastFrame();
-            counter = 0;
-        }
-        __syncthreads();
-        if (!frame.isValid())
-            return; //end kernel if we have no frames
-        filter.setSuperCellPosition((superCellIdx - mapper.getGuardingSuperCells()) * mapper.getSuperCellSize());
-        while (frame.isValid())
-        {
-            if (linearThreadIdx < particlesInSuperCell)
-            {
-                if (filter(*frame, linearThreadIdx))
-                    nvidia::atomicAllInc(&counter);
-            }
-            __syncthreads();
-            if (linearThreadIdx == 0)
+        ForEachIdx<
+            IdxConfig<
+                1,
+                numWorkers
+            >
+        > onlyMaster{ workerIdx };
+
+        onlyMaster(
+            [&](
+                uint32_t const,
+                uint32_t const
+            )
             {
-                frame = pb.getPreviousFrame(frame);
-                particlesInSuperCell = math::CT::volume<SuperCellSize>::type::value;
+                frame = pb.getLastFrame( superCellIdx );
+                particlesInSuperCell = pb.getSuperCell( superCellIdx ).getSizeLastFrame( );
+                counter = 0;
             }
-            __syncthreads();
-        }
+        );
 
-        __syncthreads();
-        if (linearThreadIdx == 0)
+        __syncthreads( );
+
+        if( !frame.isValid() )
+            return; //end kernel if we have no frames
+        filter.setSuperCellPosition(
+            ( superCellIdx - mapper.getGuardingSuperCells( ) ) *
+            mapper.getSuperCellSize( )
+        );
+
+        ForEachIdx<
+            IdxConfig<
+                frameSize,
+                numWorkers
+            >
+        > forEachParticle( workerIdx );
+
+        while( frame.isValid( ) )
         {
-            atomicAdd(gCounter, (uint64_cu) counter);
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    if( linearIdx < particlesInSuperCell )
+                    {
+                        if(
+                            filter(
```

Comment:
I would suggest keeping the old formating:
```C++
if ( filter( *frame, linearThreadIdx ) )
```
I know, this breaks some new rules, but just looks the best...